src/message.cr

```diff
@@ -18,15 +17,88 @@ module Crylog
     getter datetime : Time
 
     # Any extra metadata added by a `Crylog::Processors::LogProcessor`.
-    property extra : Hash(String, Crylog::Context)
+    setter extra : Hash(String, Crylog::Context)?
 
-    # Any extra metadata added when the message was logged.
-    getter context : Hash(String, Crylog::Context)
+    @context : Hash(String, Crylog::Context)?
 
     # Represents the formatted version of this message.
     @[JSON::Field(ignore: true)]
     property formatted : String = ""
 
-    def initialize(@message : String, @context : Hash(String, Crylog::Context), @severity : Crylog::Severity, @channel : String, @datetime : Time, @extra : Hash(String, Crylog::Context)); end
+    # Lambda held for lazy message evaluation.
+    @[JSON::Field(ignore: true)]
+    @block : -> Logger::MsgType
+
+    @[JSON::Field(ignore: true)]
+    @set_message_context = false
+
+    def initialize(@severity : Crylog::Severity, @channel : String, &block : -> Logger::MsgType)
+      @datetime = Time.utc
+      @block = block
+    end
+
+    # The message that was logged.
+    def message : String
```

Comment:
I didn't know that.

Reply:
`set_message_context` sets two variables, some of which may be `nil`.  `@context` for example stays `nil` to avoid empty object creation.   There is a measurable difference in runtime when creating empty `Hash` objects.

Would using a generator for `context` generate less efficient code by doing branch checks for `nil` which already and must be handled by `set_message_context`?

Currently calling `.context? => true` means `.context` is safe to call and `.context? => false` means `.context` may raise because it could be `nil`.

The situation is different for `.message` only because `String`s are considered immutable and `""` I think returns the same reusable object.

Alternatively I could either:
1. return a new object if `.context` is called when `.context?` is `false`
2. return a single shared object if `.context` is called when `.context?` is `false`

I'd prefer to leave it the way it is as I expect it will help catch bugs by `Formatters` misusing `context` and `context?`  It helped me catch an early one with the line formatter.